pkg/agent/controller/traceflow/traceflow_controller.go

```diff
@@ -312,144 +364,176 @@ func (c *Controller) validateTraceflow(tf *opsv1alpha1.Traceflow) error {
+				if hasInterface {
+					packet.DestinationMAC = dstPodInterface.MAC
+				}
+			}
+		} else if tf.Spec.Destination.Pod != "" {
+			dstPodInterfaces := c.interfaceStore.GetContainerInterfacesByPod(tf.Spec.Destination.Pod, tf.Spec.Destination.Namespace)
+			if len(dstPodInterfaces) > 0 {
+				if isIPv6 {
+					packet.DestinationIP = dstPodInterfaces[0].GetIPv6Addr()
+				} else {
+					packet.DestinationIP = dstPodInterfaces[0].GetIPv4Addr()
+				}
+				if !liveTraffic {
+					packet.DestinationMAC = dstPodInterfaces[0].MAC
+				}
 			} else {
-				dstIP = dstPodInterfaces[0].GetIPv4Addr().String()
+				dstPod, err := c.kubeClient.CoreV1().Pods(tf.Spec.Destination.Namespace).Get(context.TODO(), tf.Spec.Destination.Pod, metav1.GetOptions{})
+				if err != nil {
+					return nil, fmt.Errorf("failed to get the destination Pod: %v", err)
+				}
+				// DestinationMAC is nic here, will be set to Gateway
+				// MAC in ofClient.SendTraceflowPacket()
+				podIPs := make([]net.IP, len(dstPod.Status.PodIPs))
+				for i, ip := range dstPod.Status.PodIPs {
+					podIPs[i] = net.ParseIP(ip.IP)
+				}
+				if isIPv6 {
+					packet.DestinationIP, _ = util.GetIPWithFamily(podIPs, util.FamilyIPv6)
+				} else {
+					packet.DestinationIP = util.GetIPv4Addr(podIPs)
+				}
 			}
-		} else {
-			dstPod, err := c.kubeClient.CoreV1().Pods(tf.Spec.Destination.Namespace).Get(context.TODO(), tf.Spec.Destination.Pod, metav1.GetOptions{})
+			if packet.DestinationIP == nil {
+				if isIPv6 {
+					return nil, errors.New("destination Pod does not have an IPv6 address")
+				} else {
+					return nil, errors.New("destination Pod does not have an IPv4 address")
+				}
+			}
+		} else if tf.Spec.Destination.Service != "" {
+			dstSvc, err := c.serviceLister.Services(tf.Spec.Destination.Namespace).Get(tf.Spec.Destination.Service)
 			if err != nil {
-				return err
+				return nil, fmt.Errorf("failed to get the destination Service: %v", err)
 			}
-			// dstMAC is "" here, will be set to Gateway MAC in ofClient.SendTraceflowPacket
-			podIPs := make([]net.IP, len(dstPod.Status.PodIPs))
-			for i, ip := range dstPod.Status.PodIPs {
-				podIPs[i] = net.ParseIP(ip.IP)
+			if dstSvc.Spec.ClusterIP == "" {
+				return nil, errors.New("destination Service does not have a ClusterIP")
 			}
-			if isIPv6 {
-				ipv6, _ := util.GetIPWithFamily(podIPs, util.FamilyIPv6)
-				dstIP = ipv6.String()
-			} else {
-				dstIP = util.GetIPv4Addr(podIPs).String()
+			packet.DestinationIP = net.ParseIP(dstSvc.Spec.ClusterIP)
+			if !isIPv6 {
+				packet.DestinationIP = packet.DestinationIP.To4()
+				if packet.DestinationIP == nil {
+					return nil, errors.New("destination Service does not have an IPv4 ClusterIP")
+				}
+			} else if packet.DestinationIP.To4() != nil {
+				return nil, errors.New("destination Service does not have an IPv6 ClusterIP")
+			}
+
+			if !liveTraffic {
+				// Why need to set TCP flags to 2?
```

Comment:
@gran-vmv : why we need to set TCP flags to 2 for only Service? I hope to add a comment here.

Reply:
For service, it will be DNATed, thus SYN flag is required. For other traffic, it will bypass traceflowConntrachFlow.
I think we need to set the default TCP flag to 2.